### DISCUSSION TITLE
VPN / SSH Remote access

### DISCUSSION THREAD

**Discussion by zekageri** [espressif/arduino-esp32#8295] [Q&A]
Do anyone know about any solution for remote access?
My current solution is the following.

I have a server in the cloud which has a public IP and it hosts a raw websocket server.
ESP32 connects to this websocket as a client and sends every data to it.
Client can open an HTML page from the public server and the server forwards the esp32 data to the client.

I wonder if there is any other solution to directly connect my esp32 and a client ( phone ) webpage.

VPN for example.
I have found wireguard but it just not working.

My goal would be some p2p solution where the esp32 connects to my server, and the server waits for a client to connect. If a client connected, the server would attach the esp32 and the client together and it would be a transparent bridge without websockets.

Because in my current solution, every http request is going to the server (from the client) and not to my esp32. So i'm forced to do everything via websockets.

**Comment by mrengineer7777** [comment]
There's a reason the device to cloud (server) model is used by most companies.  The key here is being able to reach the target device.  Let's say the target device is in a home network.  Then you have several things going against you:
1) Home IP address can change at any time. Mitigation: use a dynamic DNS service.  https://stevessmarthomeguide.com/dynamic-dns/
2) ISPs often block email and server ports on residential service. Workaround: host at a different port number.
3) Home firewalls often block incoming traffic.  You have to manually open ports on the firewall.

In theory setting up a private tunnel into the target's local network also works.  Then all the traffic looks local.  However maintaining a VPN or Wireguard setup can be a pain.

Or you can just have the target device poll a public server for messages every 2-60s and have your clients also talk to that server.

**Reply by mrengineer7777** [reply]
I should also mention websockets are fantastically responsive but can be expensive to host on a server.  If you can tolerate some delays a persistent HTTPS connection is the way to go.  Source: experience at work.

**Reply by zekageri** [reply]
Currently there is a https websocket hosted on render with node js.

The problem with this is that I have to maintain the same frontend on the server and on the esp32. And the esp32 frontend can only talk to the esp32 via websockets because this code is the same on the server. This is so much pain too. The esp32 websocket often disconnects even on local and the esp32 must act as a socket server and client. This requires two different socket and it is really expensive in ram consumption. Expecially the https part.

**Reply by zekageri** [reply]
The thing is that i want my users to be able to reach the esp32's webserver even when there is no internet connection. That means if they want to put it on a farm with absolutely no network, they can use it in AP mode. If there is a network, they can use it in STA. But if they want to reach it from anywhere i must have a server for that. The server itself is not a problem, but the connection part. AP and STA mode can deliver the same frontend but the only solution to deliver the same frontend from anywhere is the port forwarding. But this is obviously not an ideal solution. I don't want to bother my users with such things. Currently they can register their product on my server and the server will redirect the socket messages accordingly. But i really need some p2p where every http request that goes to the server is also reflected on the esp32.

**Comment by mrengineer7777** [comment]
Consider an alternate implementation:
Write your user interface as a mobile app.  Flutter is great for Android/iOS.
On the ESP implement a REST API.
On the server implement a REST API.
Implement a mechanism to discover your devices on the local network.
Have your app discover the local devices and then automatically switch between local/cloud mode.

**Reply by zekageri** [reply]
That would be great.
But it would require to implement a communication between the server and the esp also. 
Since if the app wants to talk to the esp and it can only connect to the server, the server should be able to fetch the data from the esp32 somehow. 
And that would require a websocket connection because the server can not make requests to a device which is inside a local network.

The app would still be better because the user now has one interface to connect to both endpoints without the pain to open the server webpage or the local webpage which the esp hosts depending on the situation.

Thank you for your response.

**Comment by droscy** [comment]
Hello @zekageri

> I have found wireguard but it just not working.

The wireguard component has just been merged in [ESPHome](https://github.com/esphome/esphome/pull/4256) and it works for esp32 devices.
You can have a look at ESPHome underlying library https://github.com/droscy/esp_wireguard or at its original version https://github.com/trombik/esp_wireguard

**Reply by zekageri** [reply]
Thank you very much, i will definietly check this!

**Comment by Sarverott** [comment]
in this topic in my opinion good to mention this attempt of reverse shell: https://github.com/userlandkernel/ESPReverseTCP
